src/allmydata/test/test_istorageserver.py

```diff
@@ -0,0 +1,282 @@
+"""
+Tests for the ``IStorageServer`` interface.
+
+Note that for performance, in the future we might want the same node to be
+reused across tests, so each test should be careful to generate unique storage
+indexes.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+from future.utils import PY2, bchr
+
+if PY2:
+    # fmt: off
+    from future.builtins import filter, map, zip, ascii, chr, hex, input, next, oct, open, pow, round, super, bytes, dict, list, object, range, str, max, min  # noqa: F401
+    # fmt: on
+
+from random import randrange
+
+from twisted.internet.defer import inlineCallbacks
+from twisted.trial.unittest import TestCase
+
+from foolscap.api import Referenceable
+
+from allmydata.interfaces import IStorageServer
+from .test_system import SystemTestMixin
+
+
+def _randbytes(length):
+    # type: (int) -> bytes
+    """Return random bytes string of given length."""
+    return b"".join([bchr(randrange(0, 256)) for _ in range(length)])
+
+
+def new_storage_index():
+    # type: () -> bytes
+    """Return a new random storage index."""
+    return _randbytes(16)
+
+
+def new_secret():
+    # type: () -> bytes
+    """Return a new random secret (for lease renewal or cancellation)."""
+    return _randbytes(32)
+
+
+class IStorageServerSharedAPIsTestsMixin(object):
+    """
+    Tests for ``IStorageServer``'s shared APIs.
+
+    ``self.storage_server`` is expected to provide ``IStorageServer``.
+    """
+
+    @inlineCallbacks
+    def test_version(self):
+        # TODO get_version() returns a dict-like thing with some of the
+        # expected fields.
+        yield self.storage_server.get_version()
+
+
+class IStorageServerImmutableAPIsTestsMixin(object):
+    """
+    Tests for ``IStorageServer``'s immutable APIs.
+
+    ``self.storage_server`` is expected to provide ``IStorageServer``.
+    """
+
+    @inlineCallbacks
+    def test_allocate_buckets_new(self):
+        """
+        allocate_buckets() with a new storage index returns the matching
+        shares.
+        """
+        (already_got, allocated) = yield self.storage_server.allocate_buckets(
+            new_storage_index(),
+            new_secret(),
+            new_secret(),
+            set(range(5)),
+            1024,
+            Referenceable(),
+        )
+        self.assertEqual(already_got, set())
+        self.assertEqual(set(allocated.keys()), set(range(5)))
+        # We validate the bucket objects' interface in a later test.
+
+    @inlineCallbacks
+    def test_allocate_buckets_repeat(self):
+        """
+        allocate_buckets() with the same storage index returns the same result,
+        because the shares have not been written to.
+
+        This fails due to https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3793
```

Comment:
Elsewhere you mentioned you had figured out why it behaves this way ... something to do with how an open Foolscap connection will keep the bucket writers for incomplete shares alive and thus deny a subsequent `allocate_buckets` the chance to allocate and return them in the `allocated` mapping?

Ahh.  It seems like quite an intentional choice by the implementation, according to this snippet:

```
            elif os.path.exists(incominghome):
                # Note that we don't create BucketWriters for shnums that
                # have a partial share (in incoming/), so if a second upload
                # occurs while the first is still in progress, the second
                # uploader will use different storage servers.
```

So ... is ticket:3793 still valid?  It seems like what I told you about `allocate_buckets` (where already + allocated == requested) is at least wrong in some important ways.

Should this test change to abort or finish the shares before making the second allocate_buckets call, perhaps?

Or ... I'm beginning to get a little bit afraid that we have to think of `allocate_buckets` as the coordination protocol in a >=3 party distributed system and that's not at all what I was thinking about when I was working on GBS. :/

Reply:
There's a separate PR which changes these invariants, which I think is necessary for HTTP, so we can discuss that there: https://github.com/tahoe-lafs/tahoe-lafs/pull/1124

Reply:
I'd really like to get this merged so it's easier to review follow-up testing branches. So I suggest either switching to testing current behavior for now, or merging as is with skips.